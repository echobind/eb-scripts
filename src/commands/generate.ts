import { Command, flags } from "@oclif/command";
import { execSync } from "child_process";
import {
  getTemplateLocation,
  pathWhereScriptIsRunning,
  rootDirectory
} from "../utils/";

export const DEFAULT_COMPONENT_NAME = "MyNewComponent";
const DEFAULT_TEMPLATE_NAME = "react-component";
const DEFAULT_PATH = "src/components";

/**
 * @description valid template type
 * @example react-component
 * @todo eventually, we'll add other template types
 */
type TemplateType = "react-component";

const validTemplateTypes: TemplateType[] = ["react-component"];

export default class Generate extends Command {
  static description = "generates new files";

  static examples = [
    `$ eb-scripts generate react-component -n MyNewComponent -p src/components
    Loaded templates: _templates
    added: src/MyNewComponent.js
`
  ];

  static flags = {
    help: flags.help({ char: "h" }),
    // flag with a value (-t, --template=VALUE)
    // name should correspond with one of the following names in /_templates
    template: flags.string({
      char: "t",
      description: "template to use",
      options: validTemplateTypes
    }),
    // flag with a value (-n, --name=VALUE)
    name: flags.string({
      char: "n",
      description: "name to print",
      default: DEFAULT_COMPONENT_NAME
    }),
    // flag with a value (-p, --path=VALUE)
    path: flags.string({
      char: "p",
      default: DEFAULT_PATH,
      description: "path to where you want the files to go"
    })
  };

  static args = [
    {
      name: "templateName",
      description: "the template you want to use",
      required: true,
      options: ["react-component"] as TemplateType[]
    }
  ];

  async run() {
    const { flags, args } = this.parse(Generate);
    // Assert flags.template as a TemplateType, so TS knows it should be a TemplateType
    const template =
      (args.templateName as TemplateType) || DEFAULT_TEMPLATE_NAME;
    const name = flags.name;
    // The path where the files will go when generated
    const path = flags.path;
    const templateLocation = getTemplateLocation();
    const templatePath = `HYGEN_TMPLS=${templateLocation}`;

    this.log(
      `Generating new component using template ${template} at ${pathWhereScriptIsRunning}/${path}`
    );

    // Generate template
    execSync(
      `${templatePath} yarn gen ${template} new ${name} --root=${pathWhereScriptIsRunning} --path=${path}`,
      { cwd: rootDirectory, stdio: "inherit" }
    );
  }
}
