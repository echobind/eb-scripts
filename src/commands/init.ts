import { Command, flags } from "@oclif/command";
import { execSync } from "child_process";
import * as fs from "fs";

const DEFAULT_PROJECT_NAME = "react";

/**
 * @description a script that is used in the `package.json` scripts
 * @example "g:component": "eb-scripts generate -t react-component -n"
 */
type Scripts = {
  /* A script */
  [script in string]: string;
};

/**
 * @description valid project type
 * @example `react`
 * @todo eventually, we'll add others like `react-native`, `react-typescript`, `node`, etc.
 */
type ProjectType = "react";

/**
 * @description the available options for generate scripts that may be added to a user's `package.json`
 */
type ScriptByProject = {
  [project in ProjectType]: Scripts;
};

// Grab the path of the user's project
const pathWhereScriptIsRunning = process.cwd();
const scriptsByProject: ScriptByProject = {
  react: {
    "g:component": "eb-scripts generate -t react-component -n"
  }
};

// Use the keys from scriptsByProject as valid projec types
const validProjectTypes: ProjectType[] = Object.keys(
  scriptsByProject
) as ProjectType[];
export default class Init extends Command {
  static description =
    "initializes project by installing `eb-scripts` and adding scripts to `package.json`";

  static examples = [`$ npx eb-scripts init -p react`];

  static flags = {
    help: flags.help({ char: "h" }),
    // flag with a value (--project=VALUE)
    project: flags.string({
      description: "language/framework of project",
      options: validProjectTypes
    })
  };

  static args = [
    {
      name: "project",
      required: true,
      description: "The language or framework of the project",
      default: "react",
      options: validProjectTypes
    }
  ];

  async run() {
    const { flags, args } = this.parse(Init);
    const project = args.project || flags.project || DEFAULT_PROJECT_NAME;
    // Grab their package Json
    const packageJsonLocation = `${pathWhereScriptIsRunning}/package.json`;
    const packageJson = require(packageJsonLocation);
    const theirScripts = packageJson.scripts || {};
    const validProjectTypes = Object.keys(scriptsByProject);
    const isValidProject = validProjectTypes.includes(project);

    // Check if project is valid
    if (!isValidProject) {
      console.error(`
        Error: invalid project type.
        Please use one of the following:
        ${validProjectTypes}
        `);
      return;
    }

    this.log(`Initializing new ${project} project`);

    // Otherwise, it is valid, so we tell TypeScript to assert it as a ProjectType
    const projectScripts: Scripts = scriptsByProject[project as ProjectType];

    // Combine their scripts with our scripts
    const updatedScripts = { ...projectScripts, ...theirScripts };
    packageJson.scripts = updatedScripts;

    this.log(`Writing to package.json scripts...`);
    fs.writeFileSync(packageJsonLocation, JSON.stringify(packageJson, null, 2));

    // *******************
    // Install `eb-scripts`
    // We do this so that the scripts we added to their package.json
    // work correctly.
    // *******************

    this.log(`Installing eb-scripts as a devDependency`);

    const yarnLockPath = pathWhereScriptIsRunning + "/yarn.lock";
    const hasYarnLock = fs.existsSync(yarnLockPath);

    // Check if they're using yarn
    if (hasYarnLock) {
      execSync("yarn add --dev eb-scripts", { cwd: pathWhereScriptIsRunning });
    } else {
      execSync("npm add --save-dev eb-scripts", {
        cwd: pathWhereScriptIsRunning
      });
    }

    this.log(`Done initializing.`);
  }
}
