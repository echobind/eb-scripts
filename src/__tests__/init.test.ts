import * as path from "path";
import * as fse from "fs-extra";
import { execSync } from "child_process";
import { DEFAULT_COMPONENT_NAME } from "../commands/generate";
import { scriptsByProject } from "../commands/init";

let root = process.cwd();

describe("The `init` command", () => {
  const backupJson = "backup.json";
  const backupYarnLock = "backupYarn.lock";

  beforeAll(async () => {
    try {
      // Back up our package.json so we don't break anything during our tests
      await fse.renameSync(`${root}/package.json`, `${root}/${backupJson}`);
      // Also back up our yarn.lock
      await fse.renameSync(`${root}/yarn.lock`, `${root}/${backupYarnLock}`);
    } catch (error) {
      console.error(error);
    }
  });

  beforeEach(async () => {
    // Create a copy package.json so we can use a fresh copy for each test
    await fse.copyFile(`${root}/${backupJson}`, `${root}/package.json`);
    // Create a copy of the yarn.lock
    await fse.copyFile(`${root}/${backupYarnLock}`, `${root}/yarn.lock`);
  });

  afterEach(async () => {
    // Remove package.json and yarn.lock after each test completes
    await fse.remove(`${root}/package.json`);
    await fse.remove(`${root}/yarn.lock`);
  });

  afterAll(async () => {
    try {
      // Restore our package.json from backup
      await fse.renameSync(`${root}/${backupJson}`, `${root}/package.json`);
      // Restore our yarn lock
      await fse.renameSync(`${root}/${backupYarnLock}`, `${root}/yarn.lock`);
    } catch (err) {
      console.error(err);
    }
  });

  it("expects to find a backup.json, backupYarn.lock and package.json", async () => {
    const backupJsonExists = fse.existsSync(`${root}/${backupJson}`);
    const backupYarnLockExists = fse.existsSync(`${root}/${backupYarnLock}`);
    const packageJsonExists = fse.existsSync(`${root}/package.json`);

    expect(backupJsonExists).toBe(true);
    expect(backupYarnLockExists).toBe(true);
    expect(packageJsonExists).toBe(true);
  });

  it("throws an error if you don't pass an argument for the project type", async () => {
    // We do this because the argument of project is required
    const initCommand = () =>
      execSync(`./bin/run init`, {
        cwd: root
      });

    expect(initCommand).toThrowErrorMatchingSnapshot();
  });

  it("works with an argument of a valid project: react", async () => {
    const project = "react";
    execSync(`./bin/run init ${project}`, {
      cwd: root
    });

    const data = fse.readFileSync(`${root}/package.json`, {
      encoding: "utf8"
    });
    const packageJson = JSON.parse(data);
    const scripts = packageJson.scripts;
    console.log(scripts, "here they are");

    const devDependencies = packageJson.devDependencies;

    const hasEbScripts = Object.keys(devDependencies).includes("eb-scripts");
    const scriptName = "g:component";
    const hasGComponentScript = Object.keys(scripts).includes(scriptName);
    // Check that the g:component script matches
    const expectedScript = scriptsByProject[project]["g:component"];

    expect(hasEbScripts).toBe(true);
    expect(hasGComponentScript).toBe(true);
    expect(scripts[scriptName]).toMatch(expectedScript);
  });

  it("works with an argument of a valid project: react-typescript", async () => {
    const project = "react-typescript";

    execSync(`./bin/run init ${project}`, {
      cwd: root
    });

    const data = fse.readFileSync(`${root}/package.json`, {
      encoding: "utf8"
    });
    const packageJson = JSON.parse(data);
    const scripts = packageJson.scripts;
    console.log(scripts, "here they are");
    const devDependencies = packageJson.devDependencies;
    const hasEbScripts = Object.keys(devDependencies).includes("eb-scripts");

    const scriptName = "g:component";
    const hasGComponentScript = Object.keys(scripts).includes(scriptName);
    // Check that the g:component script matches
    const expectedScript = scriptsByProject[project]["g:component"];

    expect(hasEbScripts).toBe(true);
    expect(hasGComponentScript).toBe(true);
    expect(scripts[scriptName]).toMatch(expectedScript);
  });

  it("throws an error when you pass an invalid flag", () => {
    const initCommand = () =>
      execSync(`./bin/run init rust`, {
        cwd: root
      });

    expect(initCommand).toThrowErrorMatchingSnapshot();
  });
});
