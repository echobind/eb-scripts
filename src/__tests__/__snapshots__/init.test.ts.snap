// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The \`init\` command throws an error if you don't pass an argument for the project type 1`] = `
"Command failed: ./bin/run init
 ›   Error: Missing 1 required arg:
 ›   project  The language or framework of the project.
 ›             Valid options: react,react-typescript
 ›   See more help with --help
"
`;

exports[`The \`init\` command throws an error when you pass an invalid flag 1`] = `
"Command failed: ./bin/run init rust
 ›   Error: Expected rust to be one of: react, react-typescript
 ›   See more help with --help
"
`;
